### PR TITLE
reimplementation of Track classifiers and Track Collection mergers

### PR DESCRIPTION
**Pull request by VinInn**
have fully reimplemented track selectors and merger.
Separated all concerns:
one class to use mva,
one class to use cuts (mimics mva output)
one class to merge selectors
one class to merge collections.
to be done
re-implement duplicate identification and merging
port to HLT

do not esclude that some non-main-stream use-cases are not covered yet.
some more cleanup and optimization is surely required.

I have also implemented a macro to instantiate a sort of DynArray as C++14 does not support anymore
dynamical array.

Full backward compatibility is not maintained.
1) 
The vertex identification uses only delta-z and not the track-vertex association:
the latter (besides being utterly slow to verify with the current implementation) it is only working
for Iter0: tracks from all other iterations are not associated to any vertex
2)
quality looseSetWithPV and highPuritySetWithPV is not set (nobody is using it)
and does not make much sense when using MVA anyhow (it can mark if the mva is prompt or detached, maybe. TBD)
3) Iter5 and 6 are MVA only
4) Muon iterations cuts are not identical (and requires further tuning)
5) Duplicate rate a bit higher than before (it needs further tuning anyhow...)
6) JetCore uses a simplified cut-based selection, waiting for its own specific MVA

manually ported from CMSSW_7_5_X #9538 (original by @VinInn).

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
Some notes based on https://github.com/cms-sw/cmssw/pull/10233	 b464466  running on top of CMSSW_7_6_X_2015-07-18-2300.

There are significant physics performance changes from the reimplementation. 
(These should better be checked with a regular tracking POG validation suite).

The most notables are:
- (looks like a good thing) the number of fake displacedTracks has reduced almost to zero in ttbar events (wf 202)
wf1322_disptk_fake_n
this rate is reduced noticeably in high-pt dijets (wf 1313)
wf1313_disp_fake_pt

- there is a very large increase in the fake rate in the jetCore step without much of an change in efficiency. The effect is present in both loose and high-purity tracks, large in high purity
wf1313_core_eff_pt
wf1313_core_fake_dr
wf1313_core_fake_pt
- this is quite visible at the combined HP tracks (wf 1313)
wf1313_hp_fake_pt

 

The effect of the PR is overall somewhat more visible in high-pt dijet sample (3 TeV).
Here are some higher level comparisons for this sample.

- a few % increase in generalTracks
all_sign568vsorig_qcd13tevpt3ts3t5wf1313p0c_log10recotracks_generaltracks__reco_obj_pt
    - this appears to be mainly from jetCore iteration
all_sign568vsorig_qcd13tevpt3ts3t5wf1313p0c_recotracks_generaltracks__reco_obj_algo
- more tracks lead to more secondary vertices
all_sign568vsorig_qcd13tevpt3ts3t5wf1313p0c_recovertexs_inclusivesecondaryvertices__reco_obj_trackssize

- up to 10% increase in the number of pf charged hadrons
all_sign568vsorig_qcd13tevpt3ts3t5wf1313p0c_log10recopfcandidates_particleflow__reco_obj_pt10
- this is partially balanced by mild decrease in photons and neutral hadrons
all_sign568vsorig_qcd13tevpt3ts3t5wf1313p0c_log10recopfcandidates_particleflow__reco_obj_pt40
all_sign568vsorig_qcd13tevpt3ts3t5wf1313p0c_log10recopfcandidates_particleflow__reco_obj_pt50

- almost no change in electrons in this sample (even though the rate of gsfElectronTracks)
- the number of gsfElectronTracks  is up by up to 20% and allConversions are up by 10% both are mostly short tracks

**Comment by slava77** [comment]
I think that the fake rate in the core tracking should be fixed.
Comments or alternatives are welcome.

File: FastSimulation/Configuration/python/DataMixerPreMix_cff.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 import FastSimulation.Tracking.recoTrackAccumulator_cfi
 mixData.tracker = FastSimulation.Tracking.recoTrackAccumulator_cfi.recoTrackAccumulator.clone()
 mixData.tracker.pileUpTracks = cms.InputTag("mix","generalTracks")
-mixData.tracker.pileUpMVAValues = cms.InputTag("mix","generalTracksMVAVals")
+mixData.tracker.pileUpMVAValues = cms.InputTag("mix","generalTracksMVAValues")
```
Comment (lveldere) [inline FastSimulation/Configuration/python/DataMixerPreMix_cff.py]: @VinInn 

If you happen to update your branch before merging,
please remove this line.

**Comment by lveldere** [comment]
+1

**Comment by VinInn** [comment]
I think I will revert the selector of jet core for 760_pre2 and try to get a specific MVA for pre3

**Comment by VinInn** [comment]
@slava77 , @davidlange6 : I propose to proceed with this "tuning" for pre2.
We will work on getting something better for next prereleases

**Comment by slava77** [comment]
Vincenzo,
regarding https://github.com/cms-sw/cmssw/pull/10233#issuecomment-124117732
I'm assuming the latest commit recovers the performance.
If that's the case, the additional tuning can indeed be done in a later release.
preNext (pre2 is ~today)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by VinInn** [comment]
recovers "partially" the performance.
Depends on what the goal is

**Comment by slava77** [comment]
ok, I'll be looking for "good enough agreement" in the tests

**Comment by lveldere** [comment]
@VinInn 

just to warn you that this branch no longer merges

**Comment by slava77** [comment]
I'm wrapping up with tests in my old IB setup.
If the rebase and conflict resolution is trivial, it would not delay the integration much

**Comment by VinInn** [comment]
"rebased" as #10373

**Comment by slava77** [comment]
I'm posting here since the tests were made still with the 10233 in  CMSSW_7_6_X_2015-07-18-2300 /test area sign568/
the snapshot of the code is at slava77:CMSSW_7_6_X_2015-07-18-2300/sign568/pull10233-133c6fc

I observe reduction in tracking efficiency particularly visible in 3 TeV dijet sample with high purity selections, mainly driven by the change in performance of the jet core algorithm.
Less energetic samples are affected less.

QCD flat-pt spectrum
the most significant reduction in plain object distributions is in inclusiveSecondary Vertices: almost 10% reduction, coming primarily from 2-track vertices

all_sign568-pass-133c6fcvsorig_qcd13tevflatpt15s3000wf1338p0c_recovertexs_inclusivesecondaryvertices__reco_obj_trackssize

HP efficiency has a noticeable increase at high vtxpos, but there is also a small decrease at lower values
wf1338_hp_eff_vtxpos


In QCD 3TeV dijets
- small increase in pf charged hadrons
all_sign568-pass-133c6fcvsorig_qcd13tevpt3ts3t5wf1313p0c_log10recopfcandidates_particleflow__reco_obj_pt10
- the inclusiveSecondaryVertices content has swung backwards compared to the previous version of the code: now there is a 25% drop (there was a 10% increase)
all_sign568-pass-133c6fcvsorig_qcd13tevpt3ts3t5wf1313p0c_recovertexs_inclusivesecondaryvertices__reco_obj_trackssize

here in the 3TeV sample the jet core has also swung back to lower efficiency, which overall contributes to a small efficiency drop in HP efficiency
- track assignment partially moved to tobTec (but it's all counted as fakes in tobTec, see below)
wf1313_hp_algo
- efficiency at HP selection is down for ~prompt tracks
wf1313_hp_eff_dxy
wf1313_hp_eff_vtxpos
wf1313_hp_eff_pt

- fakes increase somewhat at higher displacements
wf1313_hp_fake_dxy
primarily on short tracks
wf1313_hp_fake_layer
wf1313_hp_fake_pt
- jetCore efficiency dropped a lot at higher displacements
wf1313_jetcorehp_eff_vtxpos
which comes with a reduction in fake rates
wf1313_jetcorehp_fake_pt
- the fakes decrease in jetcore are "offset" by an increase in tobTec
wf1313_tobtec_fake_pt

**Comment by slava77** [comment]
well, the jetcore observations are now useless, since the "rebased" as #10373 in https://github.com/cms-sw/cmssw/pull/10233#issuecomment-124856272 actually contains non-trivial changes in TrackCollectionMerger.cc and JetCoreRegionalStep_cff.py

I doubt we will look at any tracking PR soon without a complete pre-validation included in it.

**Comment by makortel** [comment]
For the record, here are tracking validation results comparing noPU, QCD 3 TeV noPU, and ttbar 25ns PU 750pre6 samples re-recoed in 760pre1 and + various options of this PR
https://mkortela.web.cern.ch/mkortela/tracking/validation/CMSSW_7_6_0_pre1_pr10233/index.html
(similar to https://github.com/cms-sw/cmssw/pull/10373#issuecomment-125512977)

**Pull request closed without merging**